<div class="container">

<table style="width: 100%;"><tr>
<td>tryStack</td>
<td style="text-align: right;">R Documentation</td>
</tr></table>
<h2>try an expression, returning the error stack</h2>

<h3>Description</h3>

<p>As in <code>try</code>, the result of an expression if it works.
If it fails, execution is not halted, but an invisible try-error class object is
returned and (unless silent=TRUE) a message <code>catted</code> to the console.<br>
Unlike <code>try</code>, <code>tryStack</code> also returns the calling stack to
trace errors and warnings and ease debugging.
</p>


<h3>Usage</h3>

<pre><code class="language-R">tryStack(
  expr,
  silent = FALSE,
  warn = TRUE,
  short = TRUE,
  file = "",
  removetry = TRUE,
  skip = NULL
)
</code></pre>


<h3>Arguments</h3>

<table>
<tr style="vertical-align: top;">
<td><code>expr</code></td>
<td>
<p>Expression to try, potentially wrapped in curly braces if
spanning several commands.</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>silent</code></td>
<td>
<p>Logical: Should printing of error message + stack be suppressed?
Does not affect warnings and messages. DEFAULT: FALSE</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>warn</code></td>
<td>
<p>Logical: trace <code>warnings</code> and <code>messages</code> also?
They are still handled like regular warnings / messages unless
<code>file !=""</code>, when they are catted into that file. DEFAULT: TRUE</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>short</code></td>
<td>
<p>Logical: should trace be abbreviated to upper -&gt; middle -&gt; lower?
If NA, it is set to TRUE for warnings and messages, FALSE for errors.
DEFAULT: TRUE</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>file</code></td>
<td>
<p>File name passed to <code>cat</code>.
If given, Errors will be appended to the file after two empty lines.
if <code>warn=T</code> and file!="", warnings and messages will not be shown,
but also appended to the file.
This is useful in lapply simulation runs.
DEFAULT: "" (catted to the console)</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>removetry</code></td>
<td>
<p>Logical: should all stack entries matching typical tryCatch
expressions be removed? Unless the call contains customized
<code>tryCatch</code> code, this can be left to the DEFAULT: TRUE</p>
</td>
</tr>
<tr style="vertical-align: top;">
<td><code>skip</code></td>
<td>
<p>Character string(s) to be removed from the stack.
e.g. "eval(expr, p)". Use short=F to find exact matches.
DEFAULT: NULL</p>
</td>
</tr>
</table>
<h3>Value</h3>

<p>Value of <code>expr</code> if evaluated successfully. If not, an invisible
object of class "try-error" as in <code>try</code> with the stack in <code>object[2]</code>.
For nested tryStack calls, <code>object[3], object[4]</code> etc. will contain "– empty error stack –"
</p>


<h3>Author(s)</h3>

<p>Berry Boessenkool, <a href="mailto:berry-b@gmx.de">berry-b@gmx.de</a>, Nov 2016
</p>


<h3>See Also</h3>

<p><code>try</code>, <code>traceCall</code>,
<a href="https://web.archive.org/web/20200813031649/https://r.789695.n4.nabble.com/Stack-trace-td4021537.html">https://web.archive.org/web/20200813031649/https://r.789695.n4.nabble.com/Stack-trace-td4021537.html</a>,
<a href="https://stackoverflow.com/questions/15282471/get-stack-trace-on-trycatched-error-in-r">https://stackoverflow.com/questions/15282471/get-stack-trace-on-trycatched-error-in-r</a>,
<a href="https://stackoverflow.com/questions/1975110/printing-stack-trace-and-continuing-after-error-occurs-in-r">https://stackoverflow.com/questions/1975110/printing-stack-trace-and-continuing-after-error-occurs-in-r</a>,
<a href="https://stackoverflow.com/questions/16879821/save-traceback-on-error-using-trycatch">https://stackoverflow.com/questions/16879821/save-traceback-on-error-using-trycatch</a>
</p>


<h3>Examples</h3>

<pre><code class="language-R">
# Functions -----

lower &lt;- function(a) {message("fake message, a = ", a); a+10}
middle &lt;- function(b) {plot(b, main=b) ; warning("fake warning, b = ", b); lower(b) }
upper &lt;- function(c) {cat("printing c:", c, "\n") ; middle(c)}
d &lt;- upper(42)
d
rm(d)


# Classical error management with try -----

is.error( d &lt;- upper("42"), TRUE, TRUE) # error, no d creation
traceback()                     # calling stack, but only in interactive mode

d &lt;- try(upper("42"), silent=TRUE)      # d created
cat(d)                                  # with error message, but no traceback
inherits(d, "try-error")                # use for coding


# way cooler with tryStack -----

d &lt;- tryStack(upper("42") ) # like try, but with traceback, even for warnings
cat(d)
d &lt;- tryStack(upper("42"), silent=TRUE, warn=0) # don't trace warnings
d &lt;- tryStack(upper("42"), short=FALSE)

tryStack(upper(42)) # returns normal output, but warnings are easier to debug
# Note: you can also set options(showWarnCalls=TRUE)

stopifnot(inherits(d, "try-error"))
stopifnot(tryStack(upper(42))==52)

## Not run:  ## file writing not wanted by CRAN checks
d &lt;- tryStack(upper("42"), silent=TRUE, file="log.txt")
openFile("log.txt")
unlink("log.txt")

## End(Not run)

op &lt;- options(warn=2)
d &lt;- try(upper("42") )
cat(d)
d &lt;- tryStack(upper("42") )
d &lt;- tryStack(upper("42"), warn=FALSE)
cat(d)
options(op) ; rm(op)

# Nested calls -----

f &lt;- function(k) tryStack(upper(k), silent=TRUE)
d &lt;- f(42)                 ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f("42")               ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- tryStack(f(4) )       ; cat("-----\n", d, "\n-----\n") ; rm(d)
# warnings in nested calls are printed twice, unless warn=0
d &lt;- tryStack(f(4), warn=0) # could also be set within 'f'

d &lt;- tryStack(f("4"))      ; cat("-----\n", d, "\n-----\n")
d[1:3] ; rm(d)
# empty stack at begin - because of tryStack in f, no real error happened in f


# Other tests -----

cat( tryStack(upper("42")) )
f &lt;- function(k) tryStack(stop("oh oh"))
d &lt;- f(42) ; cat("-----\n", d, "\n-----\n") ; rm(d) # level 4 not helpful, but OK

# stuff with base::try
f &lt;- function(k) try(upper(k), silent=TRUE)
d &lt;- f(42)     ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f("42")   ; cat("-----\n", d, "\n-----\n") ; rm(d) # regular try output

f2 &lt;- function(k) tryStack(f(k), warn=0, silent=TRUE)
d &lt;- f2(42)    ; cat("-----\n", d, "\n-----\n") ; rm(d)
d &lt;- f2("42")  ; cat("-----\n", d, "\n-----\n") ; rm(d) # try -&gt; no error.
# -&gt; Use tryCatch and you can nest those calls. note that d gets longer.


</code></pre>


</div>